### PR TITLE
Add support for dependency (validation, completion) on workspace proj…

### PR DESCRIPTION
**Pull request by vrubezhny**
…ects #77

Content Assistant on Workspace project arfifacts

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### CODE REVIEW

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java [hunk 1]
```diff
@@ -860,4 +863,36 @@ private static CompletionItem toCompletionItem(String label, String description,
...
+	
+	private void internalCollectWorkspaceArtifacts(ICompletionRequest request, Collection<ArtifactInfo> artifactInfosCollector,
+			Optional<String> groupId, Optional<String> artifactId) {
+		DOMElement parent = request.getParentElement();
+		if (parent == null || parent.getLocalName() == null) {
+			return;
+		}
+		DOMElement grandParent = parent.getParentElement();
+		if (grandParent == null || 
+				(!!PARENT_ELT.equals(grandParent.getLocalName()) &&
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:875]: `!!` seems like a bug.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java [hunk 2]
```diff
@@ -860,4 +863,36 @@ private static CompletionItem toCompletionItem(String label, String description,
 		Map<Object, Boolean> map = new ConcurrentHashMap<>();
 		return t -> map.putIfAbsent(keyExtractor.apply(t), Boolean.TRUE) == null;
 	}
+	
+	private void internalCollectWorkspaceArtifacts(ICompletionRequest request, Collection<ArtifactInfo> artifactInfosCollector,
+			Optional<String> groupId, Optional<String> artifactId) {
+		DOMElement parent = request.getParentElement();
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:869]: I don't get why line 869..879 are necessary.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 1 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java]: Yes, For sure it is

File: lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java (hunk 2 above)
Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:869]: We need to know what is the parent node for proper filtering
We need to know what is the grand parent in order to calculate the only when we're in `<parent/> `, `<dependency/>` and `<plugin/>` nodes, if we remove these lines the proposals will be calculated for any `<groupId/>`, `<artifactID/>` and `<version/>` node contents encountered anywhere iside the document, for example inside `<project/>` node.

Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:869]: OK, but shouldn't this filter be happening in the caller method? This kind of filtering doesn't seem to be specific to workspace completion, they're also true for local repo and index completion, aren't they?

Comment (vrubezhny) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java:869]: There is actually only `<parent/>`/`<plugin/>` separation in the caller method to filter the proposals, not sure if we need to replace it with any kind of filtering that will include workspace proposals.
I believe workspace proposals are to be added into the list of proposals found in local/remote/etc. repository searchers, not replacing them.
So placing parent/dependency/plugin filter here (inside `internalCollectWorkspaceArtifacts` makes `onXMLContent` method to be a bit shorter by preventing the copying of this filter among the three different cases of the caller's switch.

**Comment by mickaelistria** [comment]
OK, let's merge as it and refactor later if needed.

**Pull request merged**